mlrun/model.py

```diff
@@ -1547,12 +1596,39 @@ def _outputs_wait_for_completion(
             )
 
     def _artifact(self, key):
-        """return artifact DataItem by key"""
-        if self.status.artifacts:
-            for a in self.status.artifacts:
-                if a["metadata"]["key"] == key:
-                    return a
-        return None
+        """
+        Return the last artifact DataItem that matches the given key.
+
+        If multiple artifacts with the same key exist, return the last one in the list.
+        If there are artifacts with different tags, the method will return the one with a tag other than 'latest'
+        if available.
+        If no artifact with the given key is found, return None.
+
+        :param key: The key of the artifact to retrieve.
+        :return: The last artifact DataItem with the given key, or None if no such artifact is found.
+        """
+        if not self.status.artifacts:
+            return None
+
+        # Collect artifacts that match the key
+        matching_artifacts = [
+            artifact
+            for artifact in self.status.artifacts
+            if artifact["metadata"].get("key") == key
+        ]
+
+        if not matching_artifacts:
+            return None
+
+        # Filter out artifacts with 'latest' tag
+        non_latest_artifacts = [
+            artifact
+            for artifact in matching_artifacts
+            if artifact["metadata"].get("tag") != "latest"
+        ]
+
+        # Return the last non-'latest' artifact if available, otherwise return the last artifact
+        return (non_latest_artifacts or matching_artifacts)[-1]
```

Comment:
if both are empty list, you will experience `list index out of range` - while it might not be a real-user case - I suggest to defend the code and return the expected value in case list is empty.

Reply:
It cannot happen because I am already checking by:
```
        if not matching_artifacts:
            return None
```
that `matching_artifacts` is not an empty list

Reply:
If there are multiple tags, why do you return the last one?
and how are they sorted?

Reply:
It is not always the last one
In the case of only one tag, `status.artifacts` includes [v1, latest]. In that case, we want to return v1.
In the case of multiple tags, `status.artifacts` includes [v1, latest, v2, v3]. In that case, we need to return the last one (v3).

Reply:
> It is not always the last one In the case of only one tag, `status.artifacts` includes [v1, latest]. In that case, we want to return v1. In the case of multiple tags, `status.artifacts` includes [v1, latest, v2, v3]. In that case, we need to return the last one (v3).

comment the heuristic on the function docs. what ensures you the list is sorted by creation date ("aka recently created is always last")?

Reply:
why not return the latest if you have multiple tags? That way you get the last created one

Reply:
because this it the bug:
```
func_v1_run = project.run_function("test",params={"tag":"v1"}, local=True)
df_v1_uri = func_v1_run.outputs["df"]
func_v2_run = project.run_function("test",params={"tag":"v2"}, local=True)
mlrun.get_dataitem(df_v1_uri)
```

When df_v1_uri returns the 'latest' artifact, and then the function is run again with v2 (making the second artifact the 'latest'),
attempting to get the data item from df_v1_uri will search for the 'latest' artifact instead of v1.

Reply:
I have sorted the artifacts by creation date